templates/extras/simp.md

```diff
@@ -1,111 +1,254 @@
+and the hypothesis `this`, which is the proof of `P` which the
+`suffices` tactic adds.
 
-Recall that `simp` lemmas are almost all of the form `X = Y` or `X ↔ Y`. Hence `simp` might work
-well for such goals. However what about goals of the form `X → Y`? You could try assuming `h : X`
-and then running either `simpa using h` or `simp {contextual := tt}` to see if Lean can deduce `Y`.
+## `dsimp`
 
-## Simp options
+`dsimp` is a variant of `simp` which only uses `simp` lemmas whose
+proof is `rfl` or `iff.rfl`, i.e. lemmas of the form `A = B` or `A ↔ B`
+where `A` and `B` are equal by definition. Like `simp` it is not
+recommended to use it in the middle of a proof, however if `dsimp`
+turns your goal into `h` then `change h` might well do the same thing.
+Another common use of `dsimp` is
 
-The behaviour of `simp` can be tweaked by `simp` variants and also by passing options to the
-algorithm. A good place to start is to look at the docstring for `simp` (write `simp` in VS Code and
-hover your mouse over it to see the docstring). Here are some examples, some of which are covered by
-the docstring and some of which are not.
+```
+dsimp only
+```
 
-1) `simp only [H1, H2, H3]` uses only lemmas `H1`, `H2`, and `H3` rather than `simp`s full
-collection of lemmas. Whyever might one want to do this in practice? Because sometimes `simp`
-simplifies things too much -- it might unfold things that you wanted to keep folded, for example.
-Another reason is that using `simp only` can speed up slow `simp` calls significantly.
+which can be a very useful way of tidying up a goal. It can be safely
+used in the middle of a proof because of the `only`. This command
+tidies up lambdas (it will turn `(λ x, f x) 37` into `f 37`, and
+`{to_fun := f, ...}.to_fun` into `f`.
 
-2) `simp [-X]` stops `simp` from using lemma `X`. One could imagine using this as another solution when one finds `simp` doing more than you would like. Recall from above that `set_option trace.simplify.rewrite true` shows you exactly which lemmas `simp` is using.
+## More advanced features.
 
-3) `simp * at *`. This simplifies everything in sight. Use if life is getting complicated.
+`simp {single_pass := tt}` -- this `single_pass` is a config option,
+one of around 16 at the time of writing. One can use `single_pass` to
+avoid loops which would otherwise occur.
 
-4) `simp {single_pass := tt}` -- this `single_pass` is a config option, one of around 16 at the time of writing. One can use `single_pass` to avoid loops which would otherwise occur; for example `nat.gcd_def` is an equality with `gcd` on both the left and right hand side, so `simp [nat.gcd_def]` is risky behaviour whereas `simp [nat.gcd_def] {single_pass := tt}` is not. As you can imagine, `simp only [h] {single_pass := tt}` here makes simp behave pretty much like `rw h`.
+Searching for `structure simp_config` in the file
+`init/meta/simp_tactic.lean` in core Lean reveals other config
+options, not all of them documented.
```

Comment:
Would it be helpful to document all of them in this file? I can probably give short descriptions of all of them with some spelunking. Most of them aren't very relevant for the average user.

Reply:
That'd be great! There is a list of them at the bottom of the file, but very few of them have descriptions.